DeviceManager/DeviceHandler.py

```diff
@@ -0,0 +1,648 @@
+"""
+    Handles CRUD operations for devices, and their configuration on the
+    FIWARE backend
+"""
+
+import logging
+from flask import request
+from flask import Blueprint
+from utils import *
+from conf import CONFIG
+from BackendHandler import OrionHandler, KafkaHandler, PersistenceHandler
+from sqlalchemy.exc import IntegrityError
+
+from DatabaseModels import db, assert_device_exists, assert_template_exists, handle_consistency_exception, \
+    assert_device_relation_exists
+from DatabaseModels import DeviceTemplate, DeviceAttr, Device, DeviceTemplateMap
+from SerializationModels import *
+from TenancyManager import init_tenant_context
+
+from app import app
+
+device = Blueprint('device', __name__)
+
+LOGGER = logging.getLogger('device-manager.' + __name__)
+LOGGER.addHandler(logging.StreamHandler())
+LOGGER.setLevel(logging.INFO)
+
+
+def serialize_full_device(orm_device):
+    data = device_schema.dump(orm_device).data
+    data['attrs'] = {}
+    for template in orm_device.templates:
+        data['attrs'][template.id] = attr_list_schema.dump(template.attrs).data
+    return data
+
+
+def auto_create_template(json_payload, new_device):
+    if ('attrs' in json_payload) and (new_device.templates is None):
+        device_template = DeviceTemplate(
+            label="device.%s template" % new_device.id)
+        db.session.add(device_template)
+        new_device.templates = [device_template]
+        load_attrs(json_payload['attrs'], device_template, DeviceAttr, db)
+
+
+def parse_template_list(template_list, new_device):
+    new_device.templates = []
+    for template_id in template_list:
+        new_device.templates.append(
+            assert_template_exists(template_id, db.session))
+
+
+def find_attribute(orm_device, attr_name, attr_type):
+    """
+    Find a particular attribute in a device retrieved from database.
+    Return the attribute, if found, 'None' otherwise
+    """
+    for template_id in orm_device['attrs']:
+        for attr in orm_device['attrs'][template_id]:
+            if (attr['label'] == attr_name) and (attr['type'] == attr_type):
+                return attr
+    return None
+
+
+class DeviceHandler(object):
```

Comment:
Since we're decoupling flask (or HTTP/REST for that matter) from the "controller" itself, one could argue we'd be better off abstracting what comes from headers, body or query parameters from the calls implemented here.

For the time being, there's no need to worry about it.